drivers/pwm/pwm_ite_it8xxx2.c

```diff
@@ -0,0 +1,264 @@
+	deviation = (targe_freq / 100) + 1;
+
+	/*
+	 * Default clock source setting is 8MHz, when ITE chip is in power
+	 * saving mode, clock source 8MHz will be gated (32.768KHz won't).
+	 * So if we still need pwm output in mode, then we should set frequency
+	 * <=324Hz in board dts. Now change prescaler clock source from 8MHz to
+	 * 32.768KHz to support pwm output in mode.
+	 * NOTE: PWM output signal maximum supported frequency 324Hz comes from
+	 *       32768 / (PWM_CTRX_MIN + 1).
+	 */
+	if ((targe_freq <= 324) && (inst->PCFSR & BIT(prs_sel))) {
+		inst->PCFSR &= ~BIT(prs_sel);
+		pwm_clk_src = (uint64_t) 32768;
+	}
+
+	/*
+	 * PWM output signal frequency is
+	 * pwm_clk_src / ((CxCPRS[15:0] + 1) * (CTRx[7:0] + 1))
+	 * NOTE: 1) define CTR minimum is 100 for more precisely when
+	 *          calculate DCR
+	 *       2) CxCPRS[15:0] value 0001h results in a divisor 2
+	 *          CxCPRS[15:0] value FFFFh results in a divisor 65536
+	 *          CTRx[7:0] value 00h results in a divisor 1
+	 *          CTRx[7:0] value FFh results in a divisor 256
+	 */
+	for (ctr = 0xFF; ctr >= PWM_CTRX_MIN; ctr--) {
+		cxcprs = (((uint32_t) pwm_clk_src) / (ctr + 1) / targe_freq) - 1;
+		if (cxcprs >= 0) {
+			actual_freq = ((uint32_t) pwm_clk_src) / (ctr + 1) / (cxcprs + 1);
+			if (abs(actual_freq - targe_freq) < deviation)
+				break;
+		}
+	}
+
+	if (cxcprs > UINT16_MAX) {
+		LOG_ERR("PWM prescaler CxCPRS only support 2 bytes !");
+		return -EINVAL;
+	}
+
+	/* Set PWM prescaler clock divide and cycle time register */
+	if (prs_sel == PWM_PRESCALER_C4) {
+		inst->C4CPRS = cxcprs & 0xFF;
+		inst->C4MCPRS = (cxcprs >> 8) & 0xFF;
+		inst->CTR1 = ctr;
+	} else if (prs_sel == PWM_PRESCALER_C6) {
+		inst->C6CPRS = cxcprs & 0xFF;
+		inst->C6MCPRS = (cxcprs >> 8) & 0xFF;
+		inst->CTR2 = ctr;
+	} else if (prs_sel == PWM_PRESCALER_C7) {
+		inst->C7CPRS = cxcprs & 0xFF;
+		inst->C7MCPRS = (cxcprs >> 8) & 0xFF;
+		inst->CTR3 = ctr;
+	}
+
+	/* Set PWM channel duty cycle register */
+	*reg_dcr = (ctr * pulse_cycles) / period_cycles;
+
+	/* PWM channel clock source not gating */
+	pwm_enable(dev, 1);
+
+	LOG_DBG("clock source freq %d, targe freq %d",
+		(uint32_t) pwm_clk_src, targe_freq);
+
+	return 0;
+}
+
+static int pwm_it8xxx2_init(const struct device *dev)
+{
+	const struct pwm_it8xxx2_cfg *config = DRV_CONFIG(dev);
+	struct pwm_it8xxx2_regs *const inst = DRV_REG(dev);
+	volatile uint8_t *reg_pcssg = (uint8_t *)config->reg_pcssg;
+	int ch = config->channel;
+	int prs_sel = config->prs_sel;
+	int pcssg_shift;
+	int pcssg_mask;
+
+	/* PWM channel clock source gating before configuring */
+	pwm_enable(dev, 0);
+
+	/* Select clock source 8MHz for prescaler */
+	inst->PCFSR |= BIT(prs_sel);
+
+	/* Bit shift and mask of prescaler clock source select group register */
+	pcssg_shift = (ch % 4) * 2;
+	pcssg_mask = prs_sel << pcssg_shift;
+
+	/* Select which prescaler output to PWM channel */
+	*reg_pcssg &= ~(0x3 << pcssg_shift);
```

Comment:
prs_sel should be prevented from overflowing the bitfield.  Can you create a macro for the PCS register mask? 

For example:
#define PCS_MASK 0x3
pcssg_mask = !~(prs_sel & PCS_MASK) << pcssg_shift);
*reg_pcssg &= ~(PCS_MASK << pccssg_shift);

Reply:
Add PCSSG_MASK, but I didn't add !~ for pcssg_mask.
Because pcssg_mask is two bits value, after ! logic operates, pcssg_mask is one bit value 0b or 1b.